### ISSUE TITLE
Improper computed object key strictness inside loop

### ISSUE THREAD

**Issue by FrankBro** [microsoft/TypeScript#61377]
### 🔎 Search Terms

computed object key

### 🕗 Version & Regression Information

Tried this for the first time, tried on 5.6 and 5.8.
Tested on typescript@next, still broken.

### ⏯ Playground Link

https://www.typescriptlang.org/play/?#code/C4TwDgpgBA0hJQLxQAYCooBIDeBnYATgJYB2A5gL4oDcAULQGYCuJAxsEQPYlTAT4AKAhFacCAEwBcUAEoix4gDxwQAGij5i5AHwBKKNgr1RJfFABuAQwA2TaMgAsABigBqKACY6tPoOy0oQKgAbRQAMl5+YGkcK1sIKgBdaQAiMAJOSAIoCAIMghTVWgpdbwYxKAFrCGAoIiQoJ2o6qEUoAGZmoldXfX8gqBMzOLsABVtcAGFuM2QR6HcARjoBodr58aZcBvm3OpWg32ABfoGg0IijmOwNiaTUgHcACwQSTih0zNycvLFCgLOgQukXw11uW2mpmA9ygaQyWR++X+gKB4RB0SwNxsdhhcK+2VySKKAxKxSAA

### 💻 Code

```ts
type Key = `* ${string}`;

function test(record: Record<Key, string>) {}

const value = 40 + 2;

test({
    [`& test: ${value}`]: "proper error",
});

for (let i = 0; i < 3; i++) {
    const valuePlusConst = value + 1;
    const valuePlus = value + i;
    test({
        [`& test: ${valuePlus}`]: "why no proper error",
        [`& test: ${valuePlusConst}`]: "proper error",
        [`& test: ${value}`]: "proper error",
    })
}
```

### 🙁 Actual behavior

No error on the entry that has a computed object key which features the loop variable.

### 🙂 Expected behavior

The line should give an error, just like all the other lines.

### Additional information about the issue

Sorry, I think the title is _NOT_ great.

When making a computed object key that needs to respect a constraint, the type checking stops working when the computed object key features a constant that changes within a loop. As you can see, simply adding a const outside to a constant 1 does not create this problem. 

I've been grilling LLMs to fix this and know what's the problem. They seem to mention it's widening that's happening within a loop? Also, a lot of the suggested solutions don't even work.

Asking elsewhere, someone said this feels like a bug which is why I'm filing it.

**Comment by RyanCavanaugh** [comment]
[An isomorphic example](https://www.typescriptlang.org/play/?ts=5.4.5#code/C4TwDgpgBA0hJQLxQAYCooBIDeBnYATgJYB2A5gL4oDcAUAGYCuJAxsEQPYlTAT4AUBCCw4EAJgC4oAJWGixAHjggANFHzFyAPgCUUbFAq0RJfFABuAQwA2jaMgAsABigBqKACY6Js1dsQABVtcKRJGAFsAIwgCJAsbOzcoAFlLYAALADoCSxIxDnD+HW8uXwTA4IBhUuA4v0T3AEYS01qOSIArRrjsWih+qAB6QagAPQB+PoGAbRQAMh4+YCkceorGXCoAXSkAIgB3dIQSDigwAg5IWJiLgl2VWgo6Yxqodo6PHqn+4bHJgagswWvHwK2wayCG2qrW2e3OlxiUBuonujzoQA) that shows what's going on here is
```ts
type Key = `* ${string}`;
function test(record: Record<Key, string>) { }
const value = 40 + 2;
const valuePlus: number = value + Math.random();
const valuePlusConst = value + 1;
const obj1 = {
    // ^?
    [`& test: ${valuePlus}`]: "why no proper error",
};

const obj2 = {
    // ^?
    [`& test: ${valuePlusConst}`]: "proper error",
};
```
`obj1` has type `{ [key: string]: string }`, which isn't an excess property violation against `Record<Key, string>` since excess properties can only be triggered on concretely-known keys.

`obj2` has type `{ "& test 43": string }`, which is a concretely-known key, so it's flagged as excess

**Comment by FrankBro** [comment]
I'm surprised this is not a defect but I guess I'm very new to typescript and more used to hard-typed languages like rust or functional languages. This feels like disrespecting the contract given by the function. 

As a human, it's obvious that the contract can easily be respected, especially since the beginning of the string is static, would be understandable if it started with a template literal variable. 

I've seen some crazy stuff done with template literals and `infer`, maybe this can be rewritten to force the first part as static.

**Comment by RyanCavanaugh** [comment]
There might be some confusion here - `Record<Key, string>` is a constraint on what values property keys which match `Key` can have. It's not a constraint on what kinds of keys *overall* can be present on the object.

**Comment by MartinJohns** [comment]
> Is there a way to enforce the latter?

There is not.

https://github.com/microsoft/typescript/wiki/faq#indirect-excess-properties-are-ok

**Comment by FrankBro** [comment]
Alright. Thank you for the help here. Have a good day :)
